src/data/polynomial/basic.lean

```diff
@@ -35,20 +35,22 @@ variables [semiring R] {p q : polynomial R}
 
 instance : inhabited (polynomial R) := add_monoid_algebra.inhabited _ _
 instance : semiring (polynomial R) := add_monoid_algebra.semiring
-instance : has_scalar R (polynomial R) := add_monoid_algebra.has_scalar
-instance : semimodule R (polynomial R) := add_monoid_algebra.semimodule
+instance {S} [semiring S] [semimodule S R] : semimodule S (polynomial R) :=
+add_monoid_algebra.semimodule
 
 instance [subsingleton R] : unique (polynomial R) := add_monoid_algebra.unique
 
 @[simp] lemma support_zero : (0 : polynomial R).support = ∅ := rfl
 
 /-- `monomial s a` is the monomial `a * X^s` -/
-def monomial (n : ℕ) (a : R) : polynomial R := finsupp.single n a
+def monomial (n : ℕ) : R →ₗ[R] polynomial R := finsupp.lsingle n
```

Comment:
Any reason for preferring `R →ₗ[R] polynomial R` over `R →+ polynomial R?

Reply:
There is a projectio from `linear_map` to `add_monoid_hom` but not vice versa.

Reply:
I was under the impression that every `A →+ B` is also a `A →ₗ[A] B` in the presence of `[semiring A]`. Am I missing an obvious counter-example?

Reply:
An obvious counter-example: `complex.conj`.